src/server/cluster/cluster_family.cc

```diff
@@ -712,15 +713,56 @@ void ClusterFamily::DflyClusterMigrationFinalize(CmdArgList args, ConnectionCont
     return cntx->SendError("Migration process is not in STABLE_SYNC state");
   }
 
-  shard_set->pool()->AwaitFiberOnAll([migration](auto*) {
-    if (const auto* shard = EngineShard::tlocal(); shard)
-      migration->Finalize(shard->shard_id());
+  const auto deleted_slots = ToSlotSet(migration->GetSlotRange());
+
+  shard_set->RunBriefInParallel(
+      [](EngineShard* shard) { shard->db_slice().SetExpireAllowed(false); });
+  VLOG(2) << "Disable expiration";
+
+  server_family_->service().proactor_pool().AwaitFiberOnAll(
+      [&deleted_slots](auto*) { ServerState::tlocal()->SetMigratedSlots(deleted_slots); });
+
+  absl::Cleanup cleanup([this] {
+    server_family_->service().proactor_pool().AwaitFiberOnAll(
+        [](auto*) { ServerState::tlocal()->SetMigratedSlots({}); });
+
+    shard_set->RunBriefInParallel(
+        [](EngineShard* shard) { shard->db_slice().SetExpireAllowed(true); });
+    VLOG(2) << "Enable expiration";
   });
 
+  DispatchTracker tracker{server_family_->GetListeners(), cntx->conn(), false /* ignore paused */,
+                          true /* ignore blocked */};
+
+  auto cb = [this, &tracker, &migration, &deleted_slots](util::ProactorBase* pb) {
+    auto blocking_filter = [&deleted_slots](ArgSlice keys) {
+      bool moved = any_of(keys.begin(), keys.end(), [&](auto k) {
+        return deleted_slots.find(ClusterConfig::KeySlot(k)) == deleted_slots.end();
+      });
+      return moved ? OpStatus::KEY_MOVED : OpStatus::OK;
+    };
+
+    server_family_->CancelBlockingOnThread(blocking_filter);
+
+    if (const auto* shard = EngineShard::tlocal(); shard) {
+      // TODO add error processing to move back into STABLE_SYNC state
+      migration->Finalize(shard->shard_id());
+    }
+    tracker.TrackOnThread();
```

Comment:
DispatchTracker needs to run on the same non-preemptive operation when we change status. Scenario:

1. The dispatch above sets `SetMigratedSlots()`, so since this point in time clients can block on `AwaitIfMigrationFinalization`
2. Now some clients are blocked
3. You use the `DispatchTracker` and it waits for the clients blocked on `AwaitIfMigrationFinalization` (it can't distinguish them from a real operation)

This is the definition of "shoot oneself in the foot" 🙂 You should re-think the order of operations, because I assume we want to block operations first, then finalize, then unblock them. Currently the DispatchTracker doesn't any useful work